### ISSUE TITLE
Make it faster using Rust

### ISSUE THREAD

**Issue by ihsanturk** [ihsanturk/neuron.vim#36]
I am seeing neuron is a bit slow on some actions when you have a big zettelkästen. I am thinking about writing the basic functions (e.g. getting last zettel, creating new zettel, searching through titles) in [Rust](https://www.rust-lang.org). Because I want to edit my zettelkästen blazing fast.

Or maybe we can write some parts of the neuron itself with Rust or C to get better performance.

Or maybe neuron can implement [memoization](https://en.wikipedia.org/wiki/Memoization) to get better performance.

In this plugin, we don't need a giant JSON of zettelkasten with their relationships and dates or etc. just for getting couple of titles. We can handle that with basic unix functions just like with [this commit](https://github.com/ihsanturk/neuron.vim/commit/3a0060bc8b35b4a7fbc334ee545b166953c7a0c7) (I do not know whether or not it is faster, we need benchmarks). 

These are the thoughts that has occurred to my mind sketchily. I wonder what @srid thinks about that.

Related to #26

**Comment by srid** [comment]
> Or maybe neuron can implement memoization to get better performance.

Yes, I plan to do this. See https://github.com/srid/neuron/issues/321

Out of curiosity, how big is your zettelkasten?

**Comment by srid** [comment]
As for Rust or C, Haskell is already fast enough for this. You first need to collect evidence of what is slowing things down (by profiling), before coming up with a plan to fix it. Shake is not very good at caching/ memoization - so I suspect that's the issue, and if so - we will have to do it ourselves.

**Comment by Nadrieril** [comment]
A LSP server would keep the graph in memory, which should solve most performance problems unless the graph becomes insanely huge

**Comment by srid** [comment]
LSP is indeed the ideal approach. We still have to implement incremental patching of (in-memory) graph.

**Comment by Nadrieril** [comment]
Do we have big enough graphs that recomputing it entirely would be too slow? Since the recomputing would be entirely asynchronous from the interaction with the editor, the worst that can happen is that the editor sees an old state; recomputing would never slow the editor down.

**Comment by ihsanturk** [comment]
> Yes, I plan to do this. See srid/neuron#321

Great to see that there is a plan.
> You first need to collect evidence of what is slowing things down (by profiling), before coming up with a plan to fix it.

That is right. I will check out profiling in vim script.

> Since the recomputing would be entirely asynchronous from the interaction with the editor

Yes, I need to update this issue after fixing #26.

**Comment by srid** [comment]
I added a `--cached` option to `neuron query`. Can you give that a try @ihsanturk ? It is in this branch: https://github.com/srid/neuron/pull/327
